oshmem/mca/memheap/base/memheap_base_select.c

```diff
@@ -106,6 +108,128 @@ static size_t _memheap_size(void)
     return (size_t) memheap_align(oshmem_shmem_info_env.symmetric_heap_size);
 }
 
+static void *memheap_mmap_get(void *hint, size_t size)
+{
+    void *addr;
+
+    addr = mmap(hint, size,
+                PROT_READ | PROT_WRITE, MAP_PRIVATE | MAP_ANONYMOUS, -1, 0);
+    if (addr == MAP_FAILED) {
+        return NULL;
+    }
+
+    return addr;
+}
+
+static int memheap_exchange_base_address(size_t size, void **address)
+{
+    int nprocs = oshmem_num_procs();
+    void *base = NULL;
+    void *ptr = NULL;
+    int rc, i;
+    void **bases;
+
+    bases = calloc(nprocs, sizeof(*bases));
+    if (NULL == bases) {
+        return OSHMEM_ERROR;
+    }
+
+    if (oshmem_my_proc_id() == 0) {
+        ptr = memheap_mmap_get(NULL, size);
+        base = ptr;
+    }
+
+    rc = oshmem_shmem_bcast(&base, sizeof(base), 0);
+    if (OSHMEM_SUCCESS != rc) {
+        MEMHEAP_ERROR("Failed to exchange allocated vma for base segment "
+                      "(error %d)", rc);
+        goto out;
+    }
+
+    if (oshmem_my_proc_id() != 0) {
+        ptr = memheap_mmap_get(base, size);
+    }
+
+    MEMHEAP_VERBOSE(100, "#%d: exchange base address: base %p: %s",
+                    oshmem_my_proc_id(), base,
+                    (base == ptr)? "ok" : "unavailable");
+
+    rc = oshmem_shmem_allgather(&ptr, bases, sizeof(ptr));
+    if (OSHMEM_SUCCESS != rc) {
+        MEMHEAP_ERROR("Failed to exchange selected vma for base segment "
+                      "(error %d)", rc);
+        goto out;
+    }
```

Comment:
ok, in that case we could depend on `mca_sshmem_base_start_address` value:
1- if 0: bcast the pointer value, and any rank unable to create fails on its side, global failure
2- if UINTPTR_MAX: bcast the pointer value, allgather so that they all fallback on default value

default could be point 2-

Reply:
fixed